Index: xserver-xorg-video-intel-2.4.0/src/i830_quirks.c
===================================================================
--- xserver-xorg-video-intel-2.4.0.orig/src/i830_quirks.c	2008-08-07 15:38:34.000000000 +0300
+++ xserver-xorg-video-intel-2.4.0/src/i830_quirks.c	2008-08-07 15:39:02.000000000 +0300
@@ -246,6 +246,8 @@
 
     /* Lenovo Napa TV (use dmi)*/
     { PCI_CHIP_I945_GM, 0x17aa, SUBSYS_ANY, quirk_lenovo_tv_dmi },
+    /* Lenovo T60/R60 needs pipe A force quirk (See LP: #243405) */
+    { PCI_CHIP_I945_GM, 0x17aa, 0x201a, quirk_pipea_force },
     /* Lenovo T61 has no TV output */
     { PCI_CHIP_I965_GM, 0x17aa, 0x20b5, quirk_ignore_tv },
     /* Lenovo 3000 v200 */
