Index: xserver-xorg-video-intel-2.4.0/src/i830_display.c
===================================================================
--- xserver-xorg-video-intel-2.4.0.orig/src/i830_display.c	2008-07-23 09:14:09.000000000 +0300
+++ xserver-xorg-video-intel-2.4.0/src/i830_display.c	2008-08-07 15:39:09.000000000 +0300
@@ -1477,6 +1477,8 @@
     /* Wait for the clocks to stabilize. */
     usleep(150);
 
+    i830_update_dsparb(pScrn);
+
     OUTREG(htot_reg, (adjusted_mode->CrtcHDisplay - 1) |
 	((adjusted_mode->CrtcHTotal - 1) << 16));
     OUTREG(hblank_reg, (adjusted_mode->CrtcHBlankStart - 1) |
@@ -1510,8 +1512,6 @@
     
     i830WaitForVblank(pScrn);
 
-    i830_update_dsparb(pScrn);
-
     /* Clear any FIFO underrun status that may have occurred normally */
     OUTREG(pipestat_reg, INREG(pipestat_reg) | FIFO_UNDERRUN);
 }
