Index: xserver-xorg-video-intel-2.4.0/src/i830_dri.c
===================================================================
--- xserver-xorg-video-intel-2.4.0.orig/src/i830_dri.c	2008-07-23 10:33:58.000000000 +0300
+++ xserver-xorg-video-intel-2.4.0/src/i830_dri.c	2008-08-07 15:38:34.000000000 +0300
@@ -386,7 +386,7 @@
 
       i = 0;
       for (accum = 0; accum <= 1; accum++) {
-	 for (depth = 0; depth <= 1; depth++) {	/* and stencil */
+	 for (depth = 1; depth >= 0; depth--) {	/* and stencil */
 	    for (db = 1; db >= 0; db--) {
 	       pConfigs[i].vid = -1;
 	       pConfigs[i].class = -1;
