Index: xserver-xorg-video-intel-2.4.0/src/common.h
===================================================================
--- xserver-xorg-video-intel-2.4.0.orig/src/common.h	2008-07-23 09:14:09.000000000 +0300
+++ xserver-xorg-video-intel-2.4.0/src/common.h	2008-08-07 15:38:34.000000000 +0300
@@ -374,6 +374,8 @@
 /* chipsets require status page in non stolen memory */
 #define HWS_NEED_NONSTOLEN(pI810) (IS_GM45(pI810) || IS_G4X(pI810))
 #define SUPPORTS_INTEGRATED_HDMI(pI810) (IS_GM45(pI810) || IS_G4X(pI810))
+/* dsparb controlled by hw only */
+#define DSPARB_HWCONTROL(pI810) (IS_G4X(pI810) || IS_GM45(pI810))
 
 #define GTT_PAGE_SIZE			KB(4)
 #define ROUND_TO(x, y)			(((x) + (y) - 1) / (y) * (y))
Index: xserver-xorg-video-intel-2.4.0/src/i830_display.c
===================================================================
--- xserver-xorg-video-intel-2.4.0.orig/src/i830_display.c	2008-08-07 15:38:34.000000000 +0300
+++ xserver-xorg-video-intel-2.4.0/src/i830_display.c	2008-08-07 15:38:34.000000000 +0300
@@ -1477,7 +1477,8 @@
     /* Wait for the clocks to stabilize. */
     usleep(150);
 
-    i830_update_dsparb(pScrn);
+    if (!DSPARB_HWCONTROL(pI830))
+	i830_update_dsparb(pScrn);
 
     OUTREG(htot_reg, (adjusted_mode->CrtcHDisplay - 1) |
 	((adjusted_mode->CrtcHTotal - 1) << 16));
Index: xserver-xorg-video-intel-2.4.0/src/i830_driver.c
===================================================================
--- xserver-xorg-video-intel-2.4.0.orig/src/i830_driver.c	2008-08-07 15:38:34.000000000 +0300
+++ xserver-xorg-video-intel-2.4.0/src/i830_driver.c	2008-08-07 15:38:34.000000000 +0300
@@ -2022,7 +2022,8 @@
    }
 
    /* Save video mode information for native mode-setting. */
-   pI830->saveDSPARB = INREG(DSPARB);
+   if (!DSPARB_HWCONTROL(pI830))
+       pI830->saveDSPARB = INREG(DSPARB);
 
    pI830->saveDSPACNTR = INREG(DSPACNTR);
    pI830->savePIPEACONF = INREG(PIPEACONF);
@@ -2165,7 +2166,8 @@
    if (!IS_I830(pI830) && !IS_845G(pI830))
      OUTREG(PFIT_CONTROL, pI830->savePFIT_CONTROL);
 
-   OUTREG(DSPARB, pI830->saveDSPARB);
+   if (!DSPARB_HWCONTROL(pI830))
+       OUTREG(DSPARB, pI830->saveDSPARB);
 
    OUTREG(DSPCLK_GATE_D, pI830->saveDSPCLK_GATE_D);
    OUTREG(RENCLK_GATE_D1, pI830->saveRENCLK_GATE_D1);
@@ -2516,7 +2518,7 @@
      * (except for mode setting, where it may occur naturally).
      * Check & ack the condition.
      */
-    if (pScrn->vtSema) {
+    if (pScrn->vtSema && !DSPARB_HWCONTROL(pI830)) {
 	if (xf86_config->crtc[0]->enabled &&
 		(INREG(PIPEASTAT) & FIFO_UNDERRUN)) {
 	    xf86DrvMsg(pScrn->scrnIndex, X_ERROR, "underrun on pipe A!\n");
