Index: xserver-xorg-video-intel-2.4.0/src/i830_quirks.c
===================================================================
--- xserver-xorg-video-intel-2.4.0.orig/src/i830_quirks.c	2008-08-07 15:38:34.000000000 +0300
+++ xserver-xorg-video-intel-2.4.0/src/i830_quirks.c	2008-08-07 15:38:34.000000000 +0300
@@ -273,8 +273,6 @@
 
     /* HP Compaq 6730s has no TV output */
     { PCI_CHIP_GM45_GM, 0x103c, 0x30e8, quirk_ignore_tv },
-    /* HP Pavilion ze4944ea needs pipe A force quirk (See LP: #242389) */
-    { PCI_CHIP_GM45_GM, 0x103c, 0x3084, quirk_pipea_force },
 
     /* Thinkpad G40 needs pipe A force quirk (See LP: #240457) */
     { PCI_CHIP_I855_GM, 0x1014, 0x0543, quirk_pipea_force },
@@ -298,6 +296,8 @@
     { PCI_CHIP_I855_GM, 0x1028, 0x00c8, quirk_pipea_force },
     /* Intel 855GM hardware (See Novell Bugzilla #406123) */
     { PCI_CHIP_I855_GM, 0x10cf, 0x1215, quirk_pipea_force },
+    /* HP Pavilion ze4944ea needs pipe A force quirk (See LP: #242389) */
+    { PCI_CHIP_I855_GM, 0x103c, 0x3084, quirk_pipea_force },
 
     /* ThinkPad X40 needs pipe A force quirk */
     { PCI_CHIP_I855_GM, 0x1014, 0x0557, quirk_pipea_force },
