Index: xserver-xorg-video-intel/src/Makefile.am
===================================================================
--- xserver-xorg-video-intel.orig/src/Makefile.am
+++ xserver-xorg-video-intel/src/Makefile.am
@@ -220,3 +220,9 @@
 intel_drv_la_SOURCES += \
 	$(INTEL_XVMC_SRCS)
 endif
+
+pcidatadir = $(datadir)/xserver-xorg/pci
+pcidata_DATA = intel.ids
+
+intel.ids : common.h
+	awk '/^#define PCI_CHIP/ {print $$3 }' $(srcdir)/common.h | sed 's/^0x/8086/' > intel.ids
