--- xserver-xorg-video-intel-2.1.1/src/i830_quirks.c	2007-08-13 21:58:29.000000000 -0700
+++ xserver-xorg-video-intel-2.1.1-quirks/src/i830_quirks.c	2007-10-29 21:23:49.000000000 -0700
@@ -54,19 +54,36 @@ static void quirk_mac_mini (I830Ptr pI83
     pI830->quirk_flag |= QUIRK_IGNORE_MACMINI_LVDS;
 }
 
+/* keep this list sorted by OEM, then by chip ID */
 static i830_quirk i830_quirk_list[] = {
+    /* Aopen mini pc */
+    { PCI_CHIP_I945_GM, 0xa0a0, SUBSYS_ANY, quirk_ignore_lvds },
+
+    /* AOpen 965GM mini pc */
+    { PCI_CHIP_I965_GM, 0x8086, 0x1999, quirk_ignore_lvds },
+
+    /* Apple Mac mini has no lvds, but macbook pro does */
+    { PCI_CHIP_I945_GM, 0x8086, 0x7270, quirk_mac_mini },
+
+    /* Dell Latitude X1 */
+    { PCI_CHIP_I945_GM, 0x1028, 0x01a3, quirk_ignore_tv },
+    /* Dell XPS 1330 */
+    { PCI_CHIP_I965_GM, 0x1028, 0x0209, quirk_ignore_tv },
+
+    /* Lenovo X60s has no TV output */
+    { PCI_CHIP_I945_GM, 0x17aa, 0x201a, quirk_ignore_tv },
     /* Lenovo T61 has no TV output */
     { PCI_CHIP_I965_GM, 0x17aa, 0x20b5, quirk_ignore_tv },
-    /* Panasonic Toughbook CF-Y4 has no TV output */
-    { PCI_CHIP_I915_GM, 0x10f7, 0x8338, quirk_ignore_tv },
     /* Lenovo 3000 v200 */
     { PCI_CHIP_I965_GM, 0x17aa, 0x3c18, quirk_ignore_tv },
-    /* Dell XPS 1330 */
-    { PCI_CHIP_I965_GM, 0x1028, 0x0209, quirk_ignore_tv },
-    /* Aopen mini pc */
-    { PCI_CHIP_I945_GM, 0xa0a0, SUBSYS_ANY, quirk_ignore_lvds },
-    /* Mac mini has no lvds, but macbook pro does */
-    { PCI_CHIP_I945_GM, 0x8086, 0x7270, quirk_mac_mini },
+
+    /* Panasonic Toughbook CF-Y4 has no TV output */
+    { PCI_CHIP_I915_GM, 0x10f7, 0x8338, quirk_ignore_tv },
+    /* Panasonic Toughbook CF-Y7 has no TV output */
+    { PCI_CHIP_I965_GM, 0x10f7, 0x8338, quirk_ignore_tv },
+
+    /* Samsung Q35 has no TV output */
+    { PCI_CHIP_I945_GM, 0x144d, 0xc504, quirk_ignore_tv },
     { 0, 0, 0, NULL },
 };
 
