--- xserver-xorg-video-intel-2.2.0.90/src/i830_driver.c.orig	2008-02-19 14:18:35.000000000 -0800
+++ xserver-xorg-video-intel-2.2.0.90/src/i830_driver.c	2008-02-19 14:23:32.000000000 -0800
@@ -1455,17 +1455,22 @@ I830PreInit(ScrnInfoPtr pScrn, int flags
     * If either XAA or EXA (exclusive) is compiled in, default to it.
     * 
     * If both are compiled in, and the user didn't specify noAccel, use the
-    * config option AccelMethod to determine which to use, defaulting to EXA
-    * if none is specified, or if the string was unrecognized.
+    * config option AccelMethod to determine which to use, defaulting to XAA
+    * if none is specified, or if the string was unrecognized, except for
+    * i965, which will default to EXA since Xv video playback does not work
+    * properly with Compiz+XAA on i965.
     *
     * All this *could* go away if we removed XAA support from this driver,
     * for example. :)
     */
    if (!pI830->noAccel) {
-#ifdef I830_USE_EXA
-       pI830->useEXA = TRUE;
+#if (defined(I830_USE_EXA) && defined(I830_USE_XAA)) || !defined(I830_USE_EXA)
+       if (IS_I965G(pI830))
+           pI830->useEXA = TRUE;
+       else
+           pI830->useEXA = FALSE;
 #else
-       pI830->useEXA = FALSE;
+       pI830->useEXA = TRUE;
 #endif
 #if defined(I830_USE_XAA) && defined(I830_USE_EXA)
        int from = X_DEFAULT;
