--- xserver-xorg-video-intel-2.6.1/src/i830_quirks.c	2009-01-21 14:02:00.000000000 -0800
+++ xf86-video-intel/src/i830_quirks.c	2009-02-04 11:16:46.000000000 -0800
@@ -222,17 +222,40 @@ static void quirk_lenovo_tv_dmi (I830Ptr
 	pI830->quirk_flag |= QUIRK_IGNORE_TV;
 }
 
+static void quirk_msi_lvds_dmi (I830Ptr pI830)
+{
+   /* MSI IM-945GSE-A has no TV output, nor a LVDS connection.
+    */
+   if (!i830_dmi_data[board_name]) {
+       ErrorF("Failed to load DMI info, MSI LVDS quirk not applied.\n");
+       return;
+   }
+   if (!strncmp(i830_dmi_data[board_name],"A9830IMS",8)) {
+       pI830->quirk_flag |= QUIRK_IGNORE_LVDS;
+       pI830->quirk_flag |= QUIRK_IGNORE_TV;
+   }
+}
+
 static void quirk_ivch_dvob (I830Ptr pI830)
 {
 	pI830->quirk_flag |= QUIRK_IVCH_NEED_DVOB;
 }
 
+/* For broken hw/bios for incorrect acpi _LID state that
+   can't be fixed with customed DSDT or other way */
+static void quirk_broken_acpi_lid (I830Ptr pI830)
+{
+	pI830->quirk_flag |= QUIRK_BROKEN_ACPI_LID;
+}
+
 /* keep this list sorted by OEM, then by chip ID */
 static i830_quirk i830_quirk_list[] = {
     /* Aopen mini pc */
     { PCI_CHIP_I915_GM, 0xa0a0, SUBSYS_ANY, quirk_ignore_lvds },
     { PCI_CHIP_I945_GM, 0xa0a0, SUBSYS_ANY, quirk_ignore_lvds },
     { PCI_CHIP_I965_GM, 0xa0a0, SUBSYS_ANY, quirk_ignore_lvds },
+    { PCI_CHIP_GM45_GM, 0xa0a0, SUBSYS_ANY, quirk_ignore_lvds },
+
     { PCI_CHIP_I965_GM, 0x8086, 0x1999, quirk_ignore_lvds },
 
     /* Apple Mac mini has no lvds, but macbook pro does */
@@ -266,6 +289,9 @@ static i830_quirk i830_quirk_list[] = {
     /* Lenovo 3000 v200 */
     { PCI_CHIP_I965_GM, 0x17aa, 0x3c18, quirk_ignore_tv },
 
+    /* MSI IM-945GSE-A has no LVDS or TV (use dmi) */
+    { PCI_CHIP_I945_GME, 0x8086, 0x27ae, quirk_msi_lvds_dmi },
+
     /* Panasonic Toughbook CF-Y4 has no TV output */
     { PCI_CHIP_I915_GM, 0x10f7, 0x8338, quirk_ignore_tv },
     /* Panasonic Toughbook CF-Y7 has no TV output */
@@ -286,6 +312,8 @@ static i830_quirk i830_quirk_list[] = {
 
     /* HP Compaq nx6110 has no TV output */
     { PCI_CHIP_I915_GM, 0x103c, 0x099c, quirk_ignore_tv },
+    /* HP Compaq nx6310 has no TV output */
+    { PCI_CHIP_I945_GM, 0x103c, 0x30aa, quirk_ignore_tv },
     /* HP Compaq 6730s has no TV output */
     { PCI_CHIP_GM45_GM, 0x103c, 0x30e8, quirk_ignore_tv },
     /* HP Compaq 2730p needs pipe A force quirk (LP: #291555) */
@@ -348,6 +376,9 @@ static i830_quirk i830_quirk_list[] = {
     /* Asus Eee Box has no LVDS */
     { PCI_CHIP_I945_GME, 0x1043, 0x1252, quirk_ignore_lvds },
 
+    /* #19239: Mirrus Centrino laptop */
+    { PCI_CHIP_I915_GM, 0x1584, 0x9800, quirk_broken_acpi_lid },
+
     { 0, 0, 0, NULL },
 };
 
