diff -Nurp xserver-xorg-video-intel-2.7.0/src/i830_video.c working/src/i830_video.c
--- xserver-xorg-video-intel-2.7.0/src/i830_video.c	2009-04-08 16:59:48.000000000 -0700
+++ working/src/i830_video.c	2009-05-08 11:50:25.000000000 -0700
@@ -2392,6 +2392,7 @@ I830PutImage(ScrnInfoPtr pScrn,
                 return BadAlloc;
             if (!pPriv->textured && drm_intel_bo_pin(pPriv->buf, 4096) != 0) {
                 drm_intel_bo_unreference(pPriv->buf);
+                pPriv->buf = NULL;
                 xf86DrvMsg(pScrn->scrnIndex, X_ERROR,
                            "Failed to pin xv buffer\n");
                 return BadAlloc;
