diff -Nurp xserver-xorg-video-intel-2.6.3/src/i830_driver.c working~virtual/src/i830_driver.c
--- xserver-xorg-video-intel-2.6.3/src/i830_driver.c	2009-03-02 11:21:07.000000000 -0800
+++ working~virtual/src/i830_driver.c	2009-04-29 14:21:54.000000000 -0700
@@ -1979,6 +1979,15 @@ I830PreInit(ScrnInfoPtr pScrn, int flags
        xf86DrvMsg(pScrn->scrnIndex, X_WARNING,
 		  "VBIOS initialization failed.\n");
 
+   if (IS_I965G(pI830) && pScrn->virtualX == 0 && pScrn->virtualY == 0) {
+       xf86DrvMsg(pScrn->scrnIndex, X_INFO,
+		  "Defaulting Virtual to 2048 2048 for i965.\n");
+       pScrn->virtualX = 2048;
+       pScrn->virtualY = 2048;
+       pScrn->display->virtualX = 2048;
+       pScrn->display->virtualY = 2048;
+   }
+
    if (pI830->use_drm_mode) {
        if (!I830DrmModeInit(pScrn))
 	   return FALSE;
