Index: xserver-xorg-video-intel/src/intel_driver.c
===================================================================
--- xserver-xorg-video-intel.orig/src/intel_driver.c
+++ xserver-xorg-video-intel/src/intel_driver.c
@@ -595,7 +595,6 @@
 
 	intel->force_fallback =
 		drmCommandNone(intel->drmSubFD, DRM_I915_GEM_THROTTLE) != 0;
-	intel->use_shadow = FALSE;
 
 	/* Enable tiling by default */
 	intel->tiling = TRUE;
@@ -608,15 +607,13 @@
 			intel->tiling = FALSE;
 	}
 
-	if (xf86IsOptionSet(intel->Options, OPTION_SHADOW)) {
-		if (xf86ReturnOptValBool(intel->Options, OPTION_SHADOW, FALSE))
-			intel->force_fallback = intel->use_shadow = TRUE;
-	}
-
+	intel->use_shadow = xf86ReturnOptValBool(intel->Options, OPTION_SHADOW, IS_I8XX(intel));
 	if (intel->use_shadow) {
-		xf86DrvMsg(scrn->scrnIndex, X_CONFIG,
+		xf86DrvMsg(scrn->scrnIndex,
+			   xf86IsOptionSet(intel->Options, OPTION_SHADOW) ? X_CONFIG : X_DEFAULT,
 			   "Shadow buffer enabled,"
 			   " GPU acceleration disabled.\n");
+		intel->force_fallback = TRUE;
 	}
 
 	/* SwapBuffers delays to avoid tearing */
