Index: xserver-xorg-video-intel/src/intel_driver.c
===================================================================
--- xserver-xorg-video-intel.orig/src/intel_driver.c
+++ xserver-xorg-video-intel/src/intel_driver.c
@@ -830,9 +830,6 @@
 
 	scrn->videoRam = device->regions[fb_bar].size / 1024;
 
-	intel->force_fallback = FALSE;
-	intel->use_shadow = FALSE;
-
 	/* Enable tiling by default */
 	intel->tiling = TRUE;
 
@@ -844,13 +841,12 @@
 			intel->tiling = FALSE;
 	}
 
-	if (xf86IsOptionSet(intel->Options, OPTION_SHADOW)) {
-		if (xf86ReturnOptValBool(intel->Options, OPTION_SHADOW, FALSE))
-			intel->force_fallback = intel->use_shadow = TRUE;
-	}
+	intel->use_shadow = xf86ReturnOptValBool(intel->Options, OPTION_SHADOW, IS_I8XX(intel));
+	intel->force_fallback = intel->use_shadow;
 
 	if (intel->use_shadow) {
-		xf86DrvMsg(scrn->scrnIndex, X_CONFIG,
+		xf86DrvMsg(scrn->scrnIndex,
+			   xf86IsOptionSet(intel->Options, OPTION_SHADOW) ? X_CONFIG : X_DEFAULT,
 			   "Shadow buffer enabled,"
 			   " GPU acceleration disabled.\n");
 		intel->directRenderingType = DRI_DISABLED;
