--- a/configure.ac
+++ b/configure.ac
@@ -79,7 +79,7 @@ XORG_DRIVER_CHECK_EXT(DPMSExtension, xex
 
 # Obtain compiler/linker options for the driver dependencies
 PKG_CHECK_MODULES(XORG, [xorg-server >= 1.6 xproto fontsproto $REQUIRED_MODULES])
-PKG_CHECK_MODULES(DRM, [libdrm >= 2.4.22])
+PKG_CHECK_MODULES(DRM, [libdrm >= 2.4.21])
 PKG_CHECK_MODULES(PCIACCESS, [pciaccess >= 0.10])
 
 sdkdir=`$PKG_CONFIG --variable=sdkdir xorg-server`
